New Component - MudChat

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->
![Screenshot 2024-12-06 184959](https://github.com/user-attachments/assets/c952e6dc-e62b-4e20-a696-6071d26544ab)
![Screenshot 2024-12-06 185154](https://github.com/user-attachments/assets/d8a33584-a454-4c57-82db-4b0da104f1fa)
![Screenshot 2024-12-06 185215](https://github.com/user-attachments/assets/bc01c9d3-ff5c-4db2-8573-ffec70ede3e6)
![Screenshot 2024-12-06 185238](https://github.com/user-attachments/assets/b320a3f3-9a8f-4041-acc6-d81596939f61)

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->
Huge shoutout to @phmatray for doing a lot of the leg work.

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
resolves #9509 
resolves #4238 
resolves #4232 
resolves #9529 

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Visual  and Unit Tests

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

phmatray committed:
create MudChat component

phmatray committed:
Merge branch 'MudBlazor:dev' into features/chatbubble

phmatray committed:
Merge branch 'MudBlazor:dev' into features/chatbubble

phmatray committed:
MudChat: Add TimelineChatExample

phmatray committed:
MudChat: Fix spacing

phmatray committed:
MudChat: Fix encoding

phmatray committed:
MudChat: Add UserAttributes to root component

phmatray committed:
Merge branch 'dev' into features/chatbubble

versile2 committed:
switch chatposition start/end to left/right

versile2 committed:
Enabled RTL support, re-renamed ChatBubblePosition to Start/End

versile2 committed:
added support for multiple mud-chat-bubble inside mud-chat

versile2 committed:
Updated Examples, adding Dense, Elevation, and Variant part 1

versile2 committed:
add mud-dense and mud-square values

versile2 committed:
Added Variant/Color as parent and child override. Adjusted SCSS to match, still need image alignment and outlined :before and docs

versile2 committed:
Removed MudChatAvatar, lots of styling and doc updates

versile2 committed:
Unit Tests, Click Events, Additional Example

versile2 committed:
Merge branch 'MudBlazor:dev' into mudchatcontinued

versile2 committed:
css change

versile2 committed:
add notice to chatdocpage

phmatray commented:
A huge thanks to you.
This is an amazing job.

versile2 commented:
Should I add a hover event to make people implementing "reply" or "like" easier? Still leave it open so they can implement those features however they wish.

dennisrahmen commented:
@versile2 I would think the hover effect is there by default but can be disabled.

Then you could use the default Blazor events for hover, click and so on.

dennisrahmen commented:
In regards to this comment [custom content](https://github.com/MudBlazor/MudBlazor/pull/9529#issuecomment-2525092436)

Is there a `Style`/`Class` property if necessary sometimes there is also a `BodyStyle` property, so to make it possible to change the margins/passings inside the bubble to enable such content like mentioned in the comment.

versile2 commented:
> In regards to this comment [custom content](https://github.com/MudBlazor/MudBlazor/pull/9529#issuecomment-2525092436)
> 
> Is there a `Style`/`Class` property if necessary sometimes there is also a `BodyStyle` property, so to make it possible to change the margins/passings inside the bubble to enable such content like mentioned in the comment.

The good news is it's a Render Fragment so you can customize to your heart's content and if you don't like any of the built in styles you can even custom them using the dev tools Inspect feature. Since each MudChat can contain more than one MudChatBubble and we are trying to keep it as simple and easy to maintain as possible there is a Class and Style function for each component. Hopefully this answers your question.

henon commented:
@versile2 are you planning on making the arrow tip position customizable? 

![image](https://github.com/user-attachments/assets/9b9e9361-8088-4a2b-adbf-7d964ac7bcdd)

Something like top vs middle vs bottom vs no arrow. I don't know how hard that would be

henon commented:
Oh, and have you looked how it looks with bigger content, i.e. multiple lines of text in a single bubble?

versile2 commented:
> @versile2 are you planning on making the arrow tip position customizable?
> 
> ![image](https://private-user-images.githubusercontent.com/44090/393552319-9b9e9361-8088-4a2b-adbf-7d964ac7bcdd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2MTUzNjAsIm5iZiI6MTczMzYxNTA2MCwicGF0aCI6Ii80NDA5MC8zOTM1NTIzMTktOWI5ZTkzNjEtODA4OC00YTJiLWFkYmYtN2Q5NjRhYzdiY2RkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDIzNDQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjMDhiYjkxMDY1NzcwNWYwMmVkMDBkZGFlMTIwYTgwMzViMzc0N2E0Y2E4YzcwODU3NGVlMWVlNmUxNzc4MTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rrk7O7-un6F3AKGijIBPQQVLL9wVNgApFkl5HmhMywE)
> 
> Something like top vs middle vs bottom vs no arrow. I don't know how hard that would be

I had not planned on it, is that something you would like to see?

versile2 committed:
Update css style, add hover example, added multiline part of example, added reference to ElementReference with Unit Test

versile2 commented:
Added an example to implement Hover functionality, added an ElementReference to the MudChatBubble and fixed a style for multiline mud-chat alignment (especially for RTL) and adjusted one of the examples to use multi-line content.

![Screenshot 2024-12-07 210217](https://github.com/user-attachments/assets/85395131-dec4-4414-b9e2-065be698f47b)

versile2 committed:
whitespace

versile2 committed:
sonarqube remove comment

henon commented:
> I had not planned on it, is that something you would like to see?

As it is now, I am pretty sure that this will come up soon. As you can see from the multi-line example, the more lines you have the more disjointed the Avatar and the chat-bubble arrow become.

![image](https://github.com/user-attachments/assets/5f603852-5607-4afc-bc9b-318ee9694546)

We could put the arrow to the top to avoid having to make it customizable and approach customization later if it is really needed and someone can make a good case for it.

Pinox commented:
In the multiline example I agree with @henon. Arrow at top would be better but his previous recommendation of arrow on top , in middle and bottom is probably a good idea. Personally, in multiline I think it would look better with no arrow.

Therefor  ( top , center , bottom and none) would add a lot of flexibility in these cases.

henon commented:
@versile2 If you prefer to merge the current state nothing is stopping us from adding stuff in further PRs since this is an experimental component anyway

versile2 committed:
allow chat bubble to work outside a mudchat and center aligned avatar

versile2 committed:
Centered Avatar, added Arrow Position + tests, adjusted styles, adjusted examples and docs.

versile2 commented:
Centered Avatar across entire bubble, created Top/Middle/Bottom/None Chat Arrow Position to position arrow. Aligned arrow's to appropriate settings. Created additional field and unit tests for arrow position and added test for mudchatbubble css classes. Lots of little style adjustments. Modified HoverExample slightly. 

![Screenshot 2024-12-08 213702](https://github.com/user-attachments/assets/249b6f26-9367-438b-a207-e27fde99cabb)
![Screenshot 2024-12-08 213725](https://github.com/user-attachments/assets/8af720c6-b6ac-427e-abb2-a7e429e3d408)
![Screenshot 2024-12-08 213803](https://github.com/user-attachments/assets/a9424d74-54f3-459e-bec6-fa3ddfb1dc6b)
![Screenshot 2024-12-08 220612](https://github.com/user-attachments/assets/48209c6a-2d23-447d-8a10-8db146103cd1)

versile2 commented:
> @versile2 If you prefer to merge the current state nothing is stopping us from adding stuff in later PRs since this is an experimental component anyway

I'm fine with that just need one person to go through and make sure I don't have any code smells or anything I missed that's obvious.

versile2 committed:
son sell fixoffset for bottom/outlined

henon reviewed (approved):
LGTM

danielchalmers commented:
Are the bubbles supposed to match the background in dark mode?

Hover item stays while scrolling:

https://github.com/user-attachments/assets/b88b1cda-41f4-4f15-add9-9c2ad212f21a

Should the bubbles have a click cursor style if they're clickable instead of text selection cursor?

![image](https://github.com/user-attachments/assets/4e2345dc-e09c-4fa1-8c1a-7a18d02461a2)

Maybe the "2 hours ago" could show the actual time on hover via `title` using local format, eg "Dec 14 2024, 12:00:00 CST"?

versile2 commented:
The time is just a text field, I could certainly spruce it up to a DateTime and add localization etc, but it was meant to KISS, let the user customize whatever they want and keep us out of the deep customization as in does one user want minutes, minutes & seconds, timespan from load from now etc. 

As for the hover event example I can fix that.

danielchalmers commented:
> The time is just a text field, I could certainly spruce it up to a DateTime and add localization etc, but it was meant to KISS, let the user customize whatever they want and keep us out of the deep customization as in does one user want minutes, minutes & seconds, timespan from load from now etc.

That makes sense!

versile2 commented:
As for the color - that is a mimic of how mudalert does it. But, tbh I don't like it so changing it back to what @phmatray used initially. 

![Screenshot 2024-12-14 143409](https://github.com/user-attachments/assets/5e442c37-35e4-443a-a0d7-158ed904e08d)
![Screenshot 2024-12-14 143418](https://github.com/user-attachments/assets/4eb26693-7c2d-452c-b4e8-2a8c7ef66988)

versile2 committed:
change default color

versile2 committed:
Added Scroll Listener to drop Hovering immediately

versile2 committed:
pointer event when onclick or oncontextclick have a delegate

versile2 commented:
> Are the bubbles supposed to match the background in dark mode?
> 
> Hover item stays while scrolling:
> Screen.Recording.2024-12-14.083841.mp4
> 
> Should the bubbles have a click cursor style if they're clickable instead of text selection cursor?
> 
> ![image](https://private-user-images.githubusercontent.com/7112040/395789257-4e2345dc-e09c-4fa1-8c1a-7a18d02461a2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyMDg1MjgsIm5iZiI6MTczNDIwODIyOCwicGF0aCI6Ii83MTEyMDQwLzM5NTc4OTI1Ny00ZTIzNDVkYy1lMDljLTRmYTEtOGMxYS03YTE4ZDAyNDYxYTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTRUMjAzMDI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDFmOGI2MTBiOWE5ZTRiNWZlYmMxNGRhNTk3MDRmZDMzNWZiNzM5MGI5NDJiOTk1NmQwZWQ5YWRlNThjYjg0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.tPnQBj7gIplfEZH9sz4T7BfTeVZ0JJON-elqnSWcubE)
> 
> Maybe the "2 hours ago" could show the actual time on hover via `title` using local format, eg "Dec 14 2024, 12:00:00 CST"?

Also adjusted pointer event for select/contextselect has a delegate

Updated the Hover example to just dump the hover event if the page is scrolled. In order to use a popover that handles that stuff I'd have to include a popover with every chat bubble, dictionary to track them, etc. I wrote it as a guide to others who might want to add that functionality without adding it into the component. I'm also open to making it better, more streamlined. @Anu6is thoughts on ChatHoverExample?

versile2 committed:
proper format for scrolllistener to pass unit test

danielchalmers commented:
How about softer colors, ex `var(--mud-palette-action-default-hover)` and `var(--mud-palette-text-primary)` like the default MudButton style:

![image](https://github.com/user-attachments/assets/75b7f6ee-3c7b-44a0-9c65-9c7acdc1b285)

![image](https://github.com/user-attachments/assets/6528f460-0437-4ab7-bccb-ea582b33cc4b)

There should be a ripple and hover effect on the clickable bubbles.

If you click a reaction it seems to stay open even if you click somewhere else on the page. In general pretty buggy.

There's a slight gap here:

![image](https://github.com/user-attachments/assets/f304a96d-432c-4976-b7c0-76eb92f8374d)

danielchalmers commented:
May make sense to use restyled MudButtons for the chat bubbles so it comes with all the styles, click, hover, etc. By default it would be Disabled but enabled when clickable

danielchalmers reviewed (changes requested):
Style changes

versile2 committed:
change default style

versile2 committed:
added ripple and hover effects if the chat bubble has a delegate for onclick or oncontextclick

versile2 committed:
updated chathoverexample

versile2 committed:
bug fixes

versile2 committed:
additional fix and doc update

versile2 commented:
> How about softer colors, ex `var(--mud-palette-action-default-hover)` and `var(--mud-palette-text-primary)` like the default MudButton style:
> 
> ![image](https://private-user-images.githubusercontent.com/7112040/396201941-75b7f6ee-3c7b-44a0-9c65-9c7acdc1b285.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODk3NTUsIm5iZiI6MTczNDM4OTQ1NSwicGF0aCI6Ii83MTEyMDQwLzM5NjIwMTk0MS03NWI3ZjZlZS0zYzdiLTQ0YTAtOWM2NS05YzdhY2RjMWIyODUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1MDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTM5ZTAwMzM0YzgzYTNiMTRlN2QyOWZjNTU2YmEwM2RlZDg5YjMxNjgyOGZiNTRlODMxMWRlNTg5MGI3M2E4NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.I1k3Yn69tl8r4rDwhchfY8EO7U633c4i14ZAJZ8rlUc)
> 
> ![image](https://private-user-images.githubusercontent.com/7112040/396202027-6528f460-0437-4ab7-bccb-ea582b33cc4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODk3NTUsIm5iZiI6MTczNDM4OTQ1NSwicGF0aCI6Ii83MTEyMDQwLzM5NjIwMjAyNy02NTI4ZjQ2MC0wNDM3LTRhYjctYmNjYi1lYTU4MmIzM2NjNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1MDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTI1MzVlMzc3YjliOGZjY2E2MTYzZjM5Y2MxOTE3MGFjNTExY2FlMzBjZjkzOWFlMjExMGVhZWFmOTU3YjI1ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.8-k5RyUsUP0dCRvlGfXHG32Ay2JcTgKhOqT2PUfTvFU)
> 
> There should be a ripple and hover effect on the clickable bubbles. Not a dealbreaker and could be a follow up PR .
> 
> If you click a reaction it seems to stay open even if you click somewhere else on the page. In general pretty buggy.
> 
> There's a slight gap here:
> 
> ![image](https://private-user-images.githubusercontent.com/7112040/396199422-f304a96d-432c-4976-b7c0-76eb92f8374d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzODk3NTUsIm5iZiI6MTczNDM4OTQ1NSwicGF0aCI6Ii83MTEyMDQwLzM5NjE5OTQyMi1mMzA0YTk2ZC00MzJjLTQ5NzYtYjdjMC03NmViOTJmODM3NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTZUMjI1MDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTRmOWViOWU4YWVlMGVmYjA4YTRmNDVkOWUxN2Y4ZWIxOWQ5Y2M2ZjdlNWM5YzFkZDJkNTU0NmQyNjU0ODY5ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Bf95GCgdgExYlbxH7LY8H6-syGUc8RKMINIORfp0A7A)

Adjusted the colors to match the button, as well as added clickable styles like the button when the bubble is clickable.

Reworked the OnHoverExample to use popover and a slightly improved example. 

Fixed the padding on the hover popover and replaced the Task.Delay with a transition effect to fix the bugginess of the popover not dissapearing. 

Updated Doc page slightly, fixed a small display bug when using mudchatfooter with no content.

danielchalmers reviewed (approved):
Looks and acts great

dennisrahmen commented:
@henon may we get another preview release after this is merged? :)

henon commented:
Thanks @phmatray and @versile2 !

> @henon may we get another preview release after this is merged? :)

Sure
